core/app/models/spree/promotion/rules/user_role.rb

```diff
@@ -0,0 +1,26 @@
+module Spree
+  class Promotion
+    module Rules
+      class UserRole < PromotionRule
+        has_many :promotion_rule_roles,
+          class_name: 'Spree::PromotionRuleRole',
+          foreign_key: :promotion_rule_id,
+          dependent: :destroy
+
+        has_many :roles,
+          through: :promotion_rule_roles,
+          class_name: 'Spree::Role'
+
+        accepts_nested_attributes_for :roles
+
+        def applicable?(promotable)
+          promotable.is_a?(Spree::Order)
+        end
+
+        def eligible?(order, options = {})
+          (roles & order.user.spree_roles).any? if order.user
```

Comment:
This seems like a nice example of something that can live outside of core that only uses defined extension points & interfaces (other than possibly the tweak to `promotion_rules/create.js.erb `? Which we could put into core if it makes sense).  We could let the community try it out for a while and merge it into after it's been widely used & battle tested.

On the performance aspect:  Personally I'd prefer to use a join table for roles -> rules since it lets us use foreign keys for data integrity.  But I know promotions are often a hot spot so I can see the appeal of using a prefence.  If we used a join table then the query could be something like this, right?

```sql
select count(*)
from spree_user_roles
inner join spree_promotion_rule_roles
  on  spree_promotion_rule_roles.promotion_rule_id = <the promotion rule id>
  and spree_promotion_rule_roles.role_id = spree_user_roles.role_id
where spree_user_roles.user_id = <the user id>
```

Are we concerned that that would not be performant enough?